crates/compiler/test_syntax/src/test_helpers.rs

```diff
@@ -308,6 +317,92 @@ impl<'a> Input<'a> {
+                Parse error was: {:?}\n\n\
+                The original code was:\n\n{}\n\n\
+                The code that failed to parse:\n\n{}\n\n\
+                The original ast was:\n\n{:#?}\n\n",
+                err,
+                self.as_str(),
+                output.as_ref().as_str(),
+                actual
+            );
+        });
+
+        // Now verify that the resultant formatting is _idempotent_ - i.e. that it doesn't change again if re-formatted
+        if check_idempotency {
+            let reformatted = reparsed_ast.format_with_migrate();
+
+            if output != reformatted {
+                eprintln!("Formatting bug; formatting with --migrate flag is not stable.\nOriginal code:\n{}\n\nFormatted code:\n{}\n\nAST:\n{:#?}\n\nReparsed AST:\n{:#?}\n\n",
+                    self.as_str(),
+                    output.as_ref().as_str(),
+                    actual,
+                    reparsed_ast);
+                eprintln!("Reformatting the formatted code changed it again, as follows:\n\n");
+
+                assert_multiline_str_eq!(output.as_ref().as_str(), reformatted.as_ref().as_str());
+            }
+        }
+
+        // Check that running format WITHOUT the --migrate flag does not change anything (all
+        // migrated syntax should be supported by the formatter as is).
+        if check_idempotency {
+            let reformatted = reparsed_ast.format();
+
+            if output != reformatted {
+                eprintln!("Formatting bug; formatting with --migrate flag is not stable.\nOriginal code:\n{}\n\nFormatted code:\n{}\n\nAST:\n{:#?}\n\nReparsed AST:\n{:#?}\n\n",
+                    self.as_str(),
+                    output.as_ref().as_str(),
+                    actual,
+                    reparsed_ast);
+                eprintln!("Reformatting the formatted code changed it again, as follows:\n\n");
+
+                assert_multiline_str_eq!(output.as_ref().as_str(), reformatted.as_ref().as_str());
+            }
+        }
+
+        if let Some(expect_panic) = canonicalize_mode {
+            if expect_panic {
+                let text = self.as_str();
+                let res = std::panic::catch_unwind(|| {
+                    let new_arena = Bump::new();
+                    actual.canonicalize(&new_arena, text);
+                });
+
+                assert!(
+                    res.is_err(),
+                    "Canonicalize was expected to panic, but it did not. \
+                    If you're running test_snapshots, you may need to remove this test from \
+                    the list of tests that are expected to panic (great!), \
+                    in `fn expect_canonicalize_panics`"
+                );
+            } else {
+                // TODO grab the output here and assert things about it.
+                // For now we just make sure that it doesn't crash on this input
+                actual.canonicalize(&arena, self.as_str());
+            }
+        }
```

Comment:
Is this a useful assertion? If the code is already checked to parse the same I'm not sure what this gets us

Reply:
This was lifted as is from the existing `check_invariants` function

Reply:
Consider removing it - I don't think this gets us much, the code is already to checked to parse idempotently. If there are certain assertions to be made about the canonical AST, those should be unit tests in the canonical module